core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -225,15 +244,40 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
         // If we are looking at current channel, request a connection if possible
         if (addressIndex.isValid()
             && subchannels.get(addressIndex.getCurrentAddress()).getSubchannel() == subchannel) {
-          addressIndex.increment();
-          requestConnection();
+          if (addressIndex.increment()) {
+            cancelScheduleTask();
+            requestConnection(); // is recursive so might hit the end of the addresses
+          }
 
-          // If no addresses remaining, go into TRANSIENT_FAILURE
-          if (!addressIndex.isValid()) {
+          if (!addressIndex.isValid()
+              && (!enableHappyEyeballs || numTfSinceAccept >= addressIndex.size())) {
+            // If no addresses remaining, go into TRANSIENT_FAILURE
             helper.refreshNameResolution();
             rawConnectivityState = TRANSIENT_FAILURE;
             updateBalancingState(TRANSIENT_FAILURE,
                 new Picker(PickResult.withError(stateInfo.getStatus())));
+            break;
+          }
+        } else if (!addressIndex.isValid()) {
```

Comment:
I feel this else-if block is hard to follow and needs cleanup. Isn't this enough:
```
else if (!addressIndex.isValid() && enableHappyEyeball && ++counter) {
  refresh name resolution.
} 
```